면접 스터디 1:1 연결하는 부분 퍼블리싱 및 로직 구현

voidmatcha opened the pull request:
> resolve #105 

기능적으로는 로컬서버와의 소켓 및 WebRTC 연동과 소켓을 활용해 모든 사용자에게 이벤트를 퍼블리싱 하는 부분까지 구현하였습니다. 

퍼블리싱은 면접자 뷰, 면접관 뷰에 대해서 진행하고 있습니다. 생각보다 고려해야 할 요소들이 많아서 함께 고민해보고 작업하면 더욱 좋을 것 같습니다.

## TODO: 
- [ ] 세팅페이지 선택 시간관리 로직 상태 적용
- [x] 기태님이 작업하신 채팅 컴포넌트 추가
  - [x] 배포된 서버의 소켓 채팅과 연결
- [x] 앞, 뒤 페이지와 연동 작업
- [ ] 소켓 시그널링 서버 완성되면 붙이기 <- 채팅 연동 후 가능성을 보고 추후 stomp를 사용하게 될 수 있음

voidmatcha committed:
refactor: socket.io disconnect시 나가지지 않는 문제, 방 인원제한이 없는 문제 해결

- 간단한 트릭을 사용하여 2명이면서 유저 정보가 존재하면 disconnect된 유저 id를 sock.id로 받아와서 필터를 거쳐 제거하고 그것을 덧잎혔다.
- 추가적으로 socketToRoom에도 유저 id에 해당하는 키를 제거하였다.

voidmatcha committed:
feat: 1:1 스터디 webRTC 적용 완료 및 각 step  제작 - WIP (#105)

- 아직 배포서버와는 연결이 되지 않아서 로컬서버 기반으로 작업
- step에 단계에 대한 부분 구현과 redux 및 API, 앞 뒤 페이지와 연동하는 작업 필요

voidmatcha committed:
feat: 소켓 서버와 커뮤니케이션 하는 요소 추가 (#105)

- 두 사람 중 한명이라도 버튼을 클릭할 경우 두 사람 모두 다음으로 넘어가도록 구성

voidmatcha committed:
chore: package-lock.json 업데이트

- material ui 모듈 설치로 인해 반영된 부분

voidmatcha commented on __mock-server/server.js:
```diff
@@ -45,15 +47,30 @@ io.on('connection', (sock) => {
-    //   users[roomID] = room;
-    // }
-    // io.emit('all users', room);
+    if (
+      users[socketToRoom[sock.id]]
+      && users[socketToRoom[sock.id]].length <= 2
+    ) {
+      users[socketToRoom[sock.id]].forEach((val) => {
+        io.to(val).emit('refresh');
+      });
+      users[socketToRoom[sock.id]] = users[socketToRoom[sock.id]].filter(
+        (val) => val !== sock.id,
+      );
+    }
+    delete socketToRoom[sock.id];
```
disconnect 로직입니다. 시간을 좀 투자하긴 했는데 결국 잘 동작하도록 개선하였습니다.

SkynI25 replied:
고생많으셨습니다..ㅠㅠ

voidmatcha commented on __mock-server/server.js:
```diff
@@ -45,15 +47,30 @@ io.on('connection', (sock) => {
   });
 
-  // 1:n 일 경우 특정 사람이 disconnect시 목록갱신 부분 업데이트 필요
+  sock.on('next', () => {
+    if (
+      users[socketToRoom[sock.id]]
+      && users[socketToRoom[sock.id]].length <= 2
+    ) {
+      users[socketToRoom[sock.id]].forEach((val) => {
+        io.to(val).emit('clicked');
+      });
+    }
+  });
+
```
두 사람중 한명이 버튼을 눌렀을 경우 두 사람 모두 실행시켜야 할 필요성이 있는 기능을 퍼블리시 하는 부분입니다.

voidmatcha commented on src/components/CamView/SmallCamView.js:
```diff
@@ -1,30 +1,38 @@
+  const ref = useRef();
+
+  useEffect(() => {
+    peerRef.on('stream', (stream) => {
+      ref.current.srcObject = stream;
+      console.log('success');
+    });
+  }, []);
   return (
     <Wrapper>
-      <WrapVideo ref={peerRef} alt="peer_cam" autoPlay muted />
+      <WrapVideo ref={ref} alt="peer_cam" autoPlay muted />
```
아까 안된다고 설명드렸던 부분인데 결국 동작하게 구성해놓았습니다.

voidmatcha commented on src/pages/PeerStudyTrainPage/PeerStudyTrainPage.js:
```diff
@@ -0,0 +1,205 @@
+  }
+
+  useEffect(() => {
+    socketRef.current = io.connect('localhost:8000');
```
이부분은 시그널링 서버 이슈가 해결되는대로 수정해야 합니다.

voidmatcha commented on src/pages/PeerStudyTrainPage/PeerStudyTrainPage.js:
```diff
@@ -0,0 +1,205 @@
+          item.peer.signal(payload.signal);
+        });
+
+        socketRef.current.on('clicked', () => {
+          setStep((prev) => prev + 1);
+        });
+
+        socketRef.current.on('refresh', () => {
+          setPeers([]);
+          setStep(0);
+        });
```
서버에서 publish 된 내용을 받는 부분입니다. clicked는 프론트에서 버튼을 클릭했을 경우 서버를 경유하여 전달되고, refresh는 disconnect가 발생한 경우를 서버에서 잡아서 publish를 해줍니다.

SkynI25 replied:
과정에 대해 상세하게 설명해주셨네요! 👍

voidmatcha commented on src/pages/PeerStudyTrainPage/PeerStudyTrainPage.js:
```diff
@@ -0,0 +1,205 @@
+            <Button
+              theme={step === STEP_FIRST ? 'gray' : 'blue'}
+              text={Fixture[step].button}
+              func={() => socketRef.current.emit('next', step + 1)}
```
위에서 publish하고있는 부분에 이벤트를 전달하기 위해 사용한 부분입니다. 

한사람이 버튼을 클릭했을 때 두 사람 모두 다음 스텝으로 넘어가야하는 경우 사용합니다.

SkynI25 replied:
이러한 방식으로 면접 순서가 바뀌는군요!
PeerStudyTrainpage 컴포넌트에서 채팅화면과 관련해서 값을 전달할 수 있을거 같아요!

voidmatcha replied:
지금으로서는 채팅은 stomp로 구현되어있고 webrtc 시그널링 소켓 서버도 stomp로 마이그레이션 된다면 걷어낼 가능성이 크네요 아직 확정은 아닙니다~

voidmatcha commented on src/pages/PeerStudyTrainPage/PeerVideo.js:
```diff
@@ -0,0 +1,88 @@
+          <WrapButtonGroup>
+            <ButtonGroup />
+          </WrapButtonGroup>
+          {peer && <SamllCamView peerRef={mediaBlobUrl} />}
```
여기가 아까 언급했던 안된다고 했던 부분인데 해결하였습니다.

voidmatcha committed:
refactor: InterviewStudyEntry과 PeerStudyTrainPage 라우팅 연결 작업

- 기존에 라우팅에 포함되지 않아서 추가 후 PeerStudyTrainPage와 연결

voidmatcha committed:
feat: 각 컴포넌트에서 stomp 연결 간편화용 useSockStomp Hooks 제작(#123)

- 예은님이 제시해주신 url을 기본적으로 포함하도록 하고 room에 구
독 chat에 퍼블리시 할 수 있는 기능을 가지고 있음

voidmatcha committed:
feat: signaling socket 연결 간편화용 useSocketSignal 제작

- 예은님에게 코드를 간결하게 제공하기 위한 목적으로 제장
- 추후 stomp 통합될 예정

voidmatcha committed:
refactor: useSockStomp, useSocketSignal hooks를 적용해 코드 간결화

- hooks 를 사용하여 코드라인수를 획기적으로 줄이고 관심사를 분리함

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -0,0 +1,106 @@
+      .then((stream) => {
+        userVideo.current.srcObject = stream;
+
+        socketRef.current.emit('join room', roomId);
```
최초 접속시 소켓서버 'join room'에 roomId를 전송하는 부분입니다.

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -0,0 +1,106 @@
+
+        socketRef.current.emit('join room', roomId);
+
+        socketRef.current.on('all users', (users) => {
+          setPeers([]);
+          const getPeers = [];
+          users.forEach((userID) => {
+            const peer = createPeer(userID, socketRef.current.id, stream);
+            peersRef.current.push({
+              peerID: userID,
+              peer,
+            });
+            getPeers.push(peer);
+          });
+          setPeers(getPeers);
+        });
```
소켓서버에서 'join room'으로 받은 roomId를 바탕으로 방이 있는지 없는지

서버측에서 접속된 해당 방에 접속된 사용자가 몇명인지에 따라 각 부분을 추가하고 

'join room'을 요청한 소켓아이디에 해당하는 부분을 제외하여 usersInThisRoom을 만들어 서버측에서 'all users'에 emit 할 경우 usersInThisRoom을 전송받을 수 있도록 하는 부분입니다.

voidmatcha replied:
https://github.com/Yapp-17th/Web_2_Client/blob/feat%2F%23105/__mock-server/server.js#L17-L34

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -0,0 +1,106 @@
+          setPeers(getPeers);
+        });
+
+        socketRef.current.on('user joined', (payload) => {
+          const item = peersRef.current.find(
+            (p) => p.peerID === payload.callerID,
+          );
+          if (!item) {
+            const peer = addPeer(payload.signal, payload.callerID, stream);
+            peersRef.current.push({
+              peerID: payload.callerID,
+              peer,
+            });
+            setPeers((users) => [...users, peer]);
+          }
+        });
```
https://github.com/Yapp-17th/Web_2_Client/blob/feat%2F%23105/__mock-server/server.js#L36-L41

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -0,0 +1,106 @@
+          }
+        });
+
+        socketRef.current.on('receiving returned signal', (payload) => {
+          const item = peersRef.current.find((p) => p.peerID === payload.id);
+          item.peer.signal(payload.signal);
+        });
```
https://github.com/Yapp-17th/Web_2_Client/blob/feat%2F%23105/__mock-server/server.js#L43-L48

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -0,0 +1,106 @@
+          item.peer.signal(payload.signal);
+        });
+
+        socketRef.current.on('refresh', () => {
+          setPeers([]);
+          setStep(0);
+        });
```
disconnect가 발생한 경우 서버측에서 refresh를 emit하도록 구성하였고 이를 바탕으로 peers와 step을 리셋하는 부분입니다.

https://github.com/Yapp-17th/Web_2_Client/blob/feat%2F%23105/__mock-server/server.js#L61-L74

voidmatcha reviewed (commented):
예은님이 시그널링 서버 stomp 마이그레이션을 수월하게 할 수 있도록 코멘트를 남겨봤습니다.

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -0,0 +1,106 @@
+    });
+
+    peer.on('signal', (signal) => {
+      socketRef.current.emit('returning signal', { signal, callerID });
```
유저가 추가될경우 서버 'returning signal'에 emit

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -0,0 +1,106 @@
+    });
+
+    peer.on('signal', (signal) => {
+      socketRef.current.emit('sending signal', {
+        userToSignal,
+        callerID,
+        signal,
+      });
```
유저를 새로말들 때 서버 'sending signal'에 emit

voidmatcha committed:
refactor: match, history를 props로 받는 부분 propTypes 추가

- ReactRouterPropTypes를 사용하면 각각에 대해서 proptypes를 지정할 수 있음

voidmatcha committed:
Merge branch 'master' of https://github.com/Yapp-17th/Web_2_Client into feat/#105

voidmatcha committed:
feat: heroku 배포를 위한 Procfile 작성

- stomp 마이그레이션 실패서 소켓서버를 배포해서 사용하는 방향으로 변경

voidmatcha committed:
refactor: 혼자연습하기 style 코드 최적화 - 1920x1080

- 나름대로 화면에 제대로 렌더링 되게끔 스타일 구성 변경

voidmatcha committed:
refactor: components index.js 및 App.js 정리

- 중복으로 Import하는  컴포넌트 제가
- 제거한 컴포넌트 index.js에서 삭제

voidmatcha committed:
refactor: InterviewChat 파일명 오타 수정 및 채팅 연동가능하게끔 기능 추가

- scrollTo로 채팅이 추가됨에 따라 아래로 스크롤 따라라게끔 하는 부분 추가
- 고정된 값 실제로 채팅이 가능하도록 분리
- 채팅을 입락하고 그것을 stomp로 전송하게끔 하는 로직 추가

voidmatcha committed:
refactor: Landing 페이지 버튼 클릭시 스크롤 부드럽게 이어지도록 smooth 옵션 추가

- window.scrollTo 에 behavior로 'smooth' 옵션을 주면 스크롤이 해당 ref까지 자연스럽게 따라가도록 함

voidmatcha committed:
refactor: PeerStudyTrainPage 에서 1:1 연결과 기본적인 동작이 가능하도록 함

- [ ] 환결설정 부분
- [ ] 면접자, 면접관
- [ ] 끝나고 평가하는 부분

voidmatcha committed:
refactor: 1:1 연습 마지막 평가 모달창 추가

- 모달에서 아직 미완성인 부분 추가
- 평가모달창 1:1 스터디 마지막에 연동
- 픽스처 내용 변경

voidmatcha commented on src/pages/SelfTrainSettingPage/SelfTrainSettingPage.js:
```diff
@@ -124,3 +122,8 @@ export default function SelfTrainSettingPage({ match }) {
   );
 }
+
+SelfTrainSettingPage.propTypes = {
+  match: ReactRouterPropTypes.match.isRequired,
+  history: ReactRouterPropTypes.history.isRequired,
+};
```
match와 history에 대한 propTypes는 이렇게 추가할 수 있습니다.

voidmatcha commented on __mock-server/Procfile:
```diff
@@ -0,0 +1 @@
+web: node server.js
```
heroku 배포를 위한 Procfile 입니다.

heroku로 배포하기 위해서는 __mock-server 폴더에 해당하는 것만 해야하는데 이 경우 아래의 명령어를 사용합니다.

```bash
git push heroku $(git subtree split --prefix=__mock-server $(git symbolic-ref --short -q HEAD)):master --force
```

voidmatcha replied:
추후 클라이언트측  stomp마이그레이션 완료되면 걷어낼 부분입니다.

voidmatcha commented on src/components/index.js:
```diff
@@ -3,9 +3,9 @@ import AuthRoute from './AuthRoute';
 import Button from './Button';
```
컴포넌트에 추가되거나 삭제되는 것이 있을 경우 해당 index 파일에 추가하면 아래와 같은 형태로 추가가 가능합니다.

```jsx
import C from '@components';

...
   return (
       ...
           <C.Button />
```

voidmatcha commented on src/components/InterviewChat/InterviewChat.js:
```diff
@@ -77,46 +78,40 @@ const WrapperBottom = styled.div`
+}) {
+  const chatBoxRef = useRef();
+
+  useEffect(() => {
+    chatBoxRef.current.scrollTo({
+      top: chatBoxRef.current.scrollHeight - chatBoxRef.current.clientHeight,
+      left: 0,
+      behavior: 'smooth',
+    });
+  }, [chatData]);
```
채팅 데이터가 추가될 경우 스크롤을 맨 하단으로 옮기도록 하는 부분입니다. behavior에 smooth 옵션을 추가해서 부드럽게 하단으로 이동하도록 하였습니다.

Lavegaa replied:
최근에 과제 볼때 해당 기능 구현했는데 다음과 같이 구현하면 되는거군요! 좋은정보 감사합니다.

voidmatcha commented on src/components/ChatInput/ChatInput.js:
```diff
@@ -64,11 +66,24 @@ const SubmitBtn = styled.span`
     <Wrapper>
-      <InputField placeholder="채팅을 입력해주세요." />
-      <SubmitBtn />
+      <InputField
+        placeholder="채팅을 입력해주세요."
+        onChange={(e) => setValue(e.target.value)}
+      />
+      <SubmitBtn
+        onClick={() => {
+          onClick(value);
+          console.log(value);
+        }}
+      />
```
채팅입력값을 stomp를 통해 모두에게 전달할 수 있도록 하는 버튼 부분입니다.  stomp용 custom hooks의 handleClick을 props로 받아 사용하도록 하였습니다.

Lavegaa replied:
소켓이랑 stomp를 사용해 채팅을 구현하는건가요?

voidmatcha replied:
네 맞습니다.

voidmatcha commented on src/pages/LandingPage/LandingHeader.js:
```diff
@@ -52,7 +52,11 @@ const WrapRightInner = styled.div`
 `;
 
-const scrollToRef = (ref) => window.scrollTo(0, ref.current.offsetTop - 130);
+const scrollToRef = (ref) => window.scrollTo({
+  top: ref.current.offsetTop - 130,
+  left: 0,
+  behavior: 'smooth',
+});
```
기존에 헤더 버튼을 클릭하면 딱딱하게 스크롤이 해당  ref로 이동하도록 했던 부분을 behavior의  smooth 옵션을 줘서 스크롤이 부드럽게 따라가도록 처리하게끔 수정한 부분입니다.

voidmatcha commented on src/components/CamView/SmallCamView.js:
```diff
@@ -15,28 +14,27 @@ const Wrapper = styled.div`
+      .then((stream) => {
+        userVideo.current.srcObject = stream;
+      });
   }, []);
```
채팅 부분이 활성화 될 경우 내 캠에대한 화면이 비활성화 되고 상대방 캠이 나오는 화면에 내화면을 작게 표시하는 부분에서 다시 내 캠 화면을 가져오도록 하는 부분입니다.

voidmatcha commented on src/hooks/useSockStomp.js:
```diff
@@ -27,9 +31,17 @@ export default function useSockStomp({
           const newMessage = JSON.parse(data.body);
           console.log(newMessage);
+
+          const chatData = {
+            time: moment(new Date()).format('HH:mm A'),
+            name: newMessage.sender,
+            content: newMessage.contents,
+          };
+          setChat((prev) => [...prev, chatData]);
```
채팅 데이터에 대한 부분입니다. 객체형태로 채팅 컴포넌트에서 적용하기 편하게끔 구조화 하였습니다.

voidmatcha commented on src/hooks/useSocketSignal.js:
```diff
@@ -43,7 +43,7 @@ export default function useSocketSignal({ roomId, setStep }) {
 
   useEffect(() => {
-    socketRef.current = io.connect('localhost:8000');
+    socketRef.current = io.connect('https://witherview.herokuapp.com/');
```
현재 stomp 마이그레이션하다가 어려움이 있어 heroku에 배포한 소켓 서버를 기반으로 작업하였는데 추후 stomp 기반으로 서버로직을 클라이언트에서 처리하도록 수정하면 걷어낼 부분입니다.

voidmatcha commented on src/pages/PeerStudyTrainPage/PeerVideo.js:
```diff
@@ -59,10 +59,14 @@ export default function PeerVideo({
-            <ButtonGroup />
-          </WrapButtonGroup>
-          {peer && <SamllCamView peerRef={mediaBlobUrl} />}
+          {isTrain && (
+            <>
+              <WrapButtonGroup>
+                <ButtonGroup />
+              </WrapButtonGroup>
+              <SamllCamView />
+            </>
+          )}
```
Button쪽은 실제로 동작하도록 추가해야 합니다. 아직 작업이 완료되지 않았습니다.

voidmatcha commented on src/pages/PeerStudyTrainPage/PeerStudyTrainPage.js:
```diff
@@ -36,83 +40,100 @@ const STEP_FIRST = 0;
+                  name="최빛"
                 />
               ))}
+            {isTrain && (
+              <InterviewChat
+                height={width / 5}
+                setInterviewer
+                chatData={chat}
+                onClick={handleClick}
+              />
+            )}
```
채팅 컴포넌트를 추가하였습니다. stomp 기반으로 정상 작동합니다.

voidmatcha reviewed (commented):
마감 마지막 작업 내용까지의 커멘트입니다.

Lavegaa reviewed (commented):
고생 많으셨습니다! 배워갈점이 많네요!
